tornado/gen.py

```diff
@@ -698,108 +683,81 @@ class Runner(object):
     """
 
     def __init__(
-        self,
-        gen: "Generator[_Yieldable, Any, _T]",
-        result_future: "Future[_T]",
-        first_yielded: _Yieldable,
+        self, gen: "Generator[_Yieldable, Any, _T]", result_future: "Future[_T]"
     ) -> None:
         self.gen = gen
         self.result_future = result_future
         self.future = _null_future  # type: Union[None, Future]
         self.running = False
         self.finished = False
         self.io_loop = IOLoop.current()
-        if self.handle_yield(first_yielded):
-            gen = result_future = first_yielded = None  # type: ignore
-            self.run()
+        self.task = asyncio.ensure_future(self.run())
+
+    async def run(self) -> None:
+        "Runs the generator to completion in the context of a task"
+        while True:
+            future = self.future
+            if future is None:
+                raise Exception("No pending future")
+            if not self.future.done():  # type: ignore
+                _step = asyncio.Event()
+
+                def step(f: "Future[_T]") -> None:
+                    _step.set()
+
+                self.io_loop.add_future(self.future, step)  # type: ignore
+                await _step.wait()
```

Comment:
I would have expected this to be simply `await self.future`. We might need to create an `Event` for some edge cases (concurrent futures, or the `moment` singleton), but most of the time I think awaiting the future directly would work.